app/Domains/RevenueStreams/Actions/UpdateRevenueStream.php

```diff
@@ -30,15 +32,26 @@ public function handle(RevenueStream $revenue_stream, array $params): RevenueStr
         return $revenue_stream;
     }
 
-    public function rules(): array
+    public function rules(ActionRequest $request): array
     {
-        return [
+        $rules = [
             'id' => ['required', 'exists:revenue_streams,id'],
             'name' => ['sometimes', 'string', 'max:255'],
             'description' => ['sometimes', 'nullable', 'string'],
             'type_id' => ['sometimes', 'exists:revenue_stream_types,id'],
-            'values' => ['sometimes', 'json'],
+            'values' => ['sometimes', 'array'],
         ];
+
+        if ($request->has('type_id')) {
+            $type = RevenueStreamType::find($request->type_id);
+            if ($type) {
+                foreach ($type->properties as $property) {
+                    $rules['values.' . $property['name']] = ['required', Rule::in(['single_line', 'multi_line', 'text', 'range', 'radio', 'checkbox', 'dropbox', 'repeater'])];
```

Comment:
Use the `RevenueStreamTypeEnum` here.
Also, please check the [comment on the `RevenueStreamType` PR](https://github.com/Eveneer/evengoams-backend/pull/62#pullrequestreview-2228760683).